# Classification Analysis of Parkinson Speech Dataset 

In this project we are predicting Parkinsons Disease by using a publicly available dataset which is splited in train and test dataset.

In the training dataset, their is an extra column which is numerical value of Parkinson Disease level which is suggested by physians so that the dataset can be used for regression also. But in our project we are just trying to classify the parkinsons. So, we dont need that column in this project from the dataset. 

# Organizing out the Dataset

The dataset is not organized properly. For our requirements, we are just organizing it with some column names.

In [29]:
import pandas as pd
columns = ["id"]
features = ["feature_"+str(i+1) for i in range(27)]
label = ["label"]

columns  = columns+features+label
df_train = pd.read_csv("train_data.txt", names = columns)
columns = columns[:27]+columns[28:]

df_test = pd.read_csv("test_data.txt", names = columns)
display(df_train.head())
display(df_test.head())

id  feature_1  feature_2  feature_3  feature_4  feature_5  feature_6  \
0   1      1.488   0.000090      0.900      0.794      2.699      8.334   
1   1      0.728   0.000038      0.353      0.376      1.059      5.864   
2   1      1.220   0.000074      0.732      0.670      2.196      8.719   
3   1      2.502   0.000123      1.156      1.634      3.469     13.513   
4   1      3.509   0.000167      1.715      1.539      5.145      9.112   

   feature_7  feature_8  feature_9  ...  feature_19  feature_20  feature_21  \
0      0.779      4.517      4.609  ...     187.576         160         159   
1      0.642      2.058      3.180  ...     234.505         170         169   
2      0.875      4.347      5.166  ...     211.442        1431        1427   
3      1.273      5.263      8.771  ...     220.230          94          92   
4      1.040      3.102      4.927  ...     225.162         117         114   

   feature_22  feature_23  feature_24  feature_25  feature_26  feature_27  \
0    0.006065    0.000416       0.000           0       0.000          23   
1    0.005181    0.000403       2.247           0       0.000          23   
2    0.006071    0.000474      10.656           1       0.178          23   
3    0.004910    0.000320       0.000           0       0.000          23   
4    0.004757    0.000380      18.182           1      13.318          23   

   label  
0      1  
1      1  
2      1  
3      1  
4      1  

[5 rows x 29 columns]

id  feature_1  feature_2  feature_3  feature_4  feature_5  feature_6  \
0   1      0.135   0.000007      0.067      0.078      0.202      2.033   
1   1      0.143   0.000007      0.073      0.081      0.219      1.236   
2   1      0.162   0.000008      0.087      0.089      0.260      1.338   
3   1      0.140   0.000007      0.075      0.089      0.224      1.086   
4   1      0.150   0.000007      0.080      0.097      0.240      1.049   

   feature_7  feature_8  feature_9  ...  feature_18  feature_19  feature_20  \
0      0.178      1.074      1.336  ...     184.502     187.880       183.0   
1      0.107      0.612      0.904  ...     198.665     202.214       194.0   
2      0.117      0.630      0.948  ...     197.220     206.060       198.0   
3      0.094      0.556      0.747  ...     202.324     206.182       200.0   
4      0.091      0.533      0.698  ...     205.407     209.927       204.0   

   feature_21  feature_22  feature_23  feature_24  feature_25  feature_26  \
0       182.0    0.005368    0.000025         0.0         0.0         0.0   
1       193.0    0.004988    0.000020         0.0         0.0         0.0   
2       197.0    0.004940    0.000046         0.0         0.0         0.0   
3       199.0    0.004900    0.000023         0.0         0.0         0.0   
4       203.0    0.004820    0.000022         0.0         0.0         0.0   

   label  
0      1  
1      1  
2      1  
3      1  
4      1  

[5 rows x 28 columns]

Dropping the column which contain numerical value to perform regression.

In [30]:
df_train.drop('feature_27', axis=1, inplace=True)

In [31]:
df_train.head()

id  feature_1  feature_2  feature_3  feature_4  feature_5  feature_6  \
0   1      1.488   0.000090      0.900      0.794      2.699      8.334   
1   1      0.728   0.000038      0.353      0.376      1.059      5.864   
2   1      1.220   0.000074      0.732      0.670      2.196      8.719   
3   1      2.502   0.000123      1.156      1.634      3.469     13.513   
4   1      3.509   0.000167      1.715      1.539      5.145      9.112   

   feature_7  feature_8  feature_9  ...  feature_18  feature_19  feature_20  \
0      0.779      4.517      4.609  ...     142.229     187.576         160   
1      0.642      2.058      3.180  ...     159.515     234.505         170   
2      0.875      4.347      5.166  ...     146.445     211.442        1431   
3      1.273      5.263      8.771  ...     182.713     220.230          94   
4      1.040      3.102      4.927  ...     182.821     225.162         117   

   feature_21  feature_22  feature_23  feature_24  feature_25  feature_26  \
0         159    0.006065    0.000416       0.000           0       0.000   
1         169    0.005181    0.000403       2.247           0       0.000   
2        1427    0.006071    0.000474      10.656           1       0.178   
3          92    0.004910    0.000320       0.000           0       0.000   
4         114    0.004757    0.000380      18.182           1      13.318   

   label  
0      1  
1      1  
2      1  
3      1  
4      1  

[5 rows x 28 columns]

Features Scaling and Setting the train, test data for training purpose and predicting.

In [32]:
predictors_train = df_train.drop(['id'], axis = 1)
predictors_test = df_test.drop(['id'], axis = 1)


predictors_train = predictors_train.drop(['label'], axis = 1).values
predictors_test = predictors_test.drop(['label'], axis = 1).values

target_train = df_train['label']
target_test = df_test['label']

from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler((-1, 1))
X_train = scaler.fit_transform(predictors_train)
X_test = scaler.fit_transform(predictors_test)

Y_train = target_train
Y_test = target_test

In [21]:
X_train.shape, X_test.shape

((1040, 26), (168, 26))

# XGBoost Algorithm

In [22]:
import xgboost as xgb
from xgboost import XGBClassifier
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn import metrics
model = XGBClassifier()
model.fit(X_train, Y_train)

y_pred = model.predict(X_test)
print("XGB boost:")
print(metrics.accuracy_score(Y_test, y_pred))
print(metrics.classification_report(Y_test, y_pred))
print(metrics.confusion_matrix(Y_test, y_pred))

[11:59:17] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.4.0/src/learner.cc:1095: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
XGB boost:
0.6845238095238095
              precision    recall  f1-score   support

           0       0.00      0.00      0.00         0
           1       1.00      0.68      0.81       168

    accuracy                           0.68       168
   macro avg       0.50      0.34      0.41       168
weighted avg       1.00      0.68      0.81       168

[[  0   0]
 [ 53 115]]


C:\Users\Nahiyan Bin Noor\anaconda3\lib\site-packages\xgboost\sklearn.py:1146: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)
C:\Users\Nahiyan Bin Noor\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\Nahiyan Bin Noor\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_

With XGBoost the accuracy is 68.45%. In case of predictiong there are zero accuracy while predicting non disease because in test data all of the instances from the persons who have parkinsons disease.

# Logistic Regression

In [33]:
from sklearn import datasets
from sklearn import metrics
from sklearn.linear_model import LogisticRegression

# fit a logistic regression model to the data
model = LogisticRegression()
model.fit(X_train, Y_train)
y_pred = model.predict(X_test)
# summarize the fit of the model
print("Logistic Regression: ")
print(metrics.accuracy_score(Y_test, y_pred))
print(metrics.classification_report(Y_test, y_pred))
print(metrics.confusion_matrix(Y_test, y_pred))

Logistic Regression: 
0.9166666666666666
              precision    recall  f1-score   support

           0       0.00      0.00      0.00         0
           1       1.00      0.92      0.96       168

    accuracy                           0.92       168
   macro avg       0.50      0.46      0.48       168
weighted avg       1.00      0.92      0.96       168

[[  0   0]
 [ 14 154]]


C:\Users\Nahiyan Bin Noor\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\Nahiyan Bin Noor\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\Nahiyan Bin Noor\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


With Logistic Regression the accuracy is 91.67%

# Gaussian Naive Bayes

In [24]:
from sklearn import datasets
from sklearn import metrics
from sklearn.naive_bayes import GaussianNB
model = GaussianNB()
model.fit(X_train, Y_train)
# make predictions
y_pred = model.predict(X_test)
# summarize the fit of the model
print("Gaussian Naive Bayes:")
print(metrics.accuracy_score(Y_test, y_pred))
print(metrics.classification_report(Y_test, y_pred))
print(metrics.confusion_matrix(Y_test, y_pred))

Gaussian Naive Bayes:
0.8511904761904762
              precision    recall  f1-score   support

           0       0.00      0.00      0.00         0
           1       1.00      0.85      0.92       168

    accuracy                           0.85       168
   macro avg       0.50      0.43      0.46       168
weighted avg       1.00      0.85      0.92       168

[[  0   0]
 [ 25 143]]


C:\Users\Nahiyan Bin Noor\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\Nahiyan Bin Noor\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\Nahiyan Bin Noor\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


With Gaussian Naive Bayes the accuracy is 85.12%

# k-Nearest Neighbor

In [25]:
from sklearn import datasets
from sklearn import metrics
from sklearn.neighbors import KNeighborsClassifier
model = KNeighborsClassifier()
model.fit(X_train, Y_train)
# make predictions
y_pred = model.predict(X_test)
# summarize the fit of the model
print("k-Nearest Neighbor: ")
print(metrics.accuracy_score(Y_test, y_pred))
print(metrics.classification_report(Y_test, y_pred))
print(metrics.confusion_matrix(Y_test, y_pred))

k-Nearest Neighbor: 
0.8392857142857143
              precision    recall  f1-score   support

           0       0.00      0.00      0.00         0
           1       1.00      0.84      0.91       168

    accuracy                           0.84       168
   macro avg       0.50      0.42      0.46       168
weighted avg       1.00      0.84      0.91       168

[[  0   0]
 [ 27 141]]


C:\Users\Nahiyan Bin Noor\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\Nahiyan Bin Noor\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\Nahiyan Bin Noor\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


With KNN the accuracy is 83.92%

# Support Vector Machine

In [26]:
from sklearn import datasets
from sklearn import metrics
from sklearn.svm import SVC
model = SVC()
model.fit(X_train, Y_train)

# make predictions
y_pred = model.predict(X_test)
# summarize the fit of the model
print("Support Vector Machine: ")
print(metrics.accuracy_score(Y_test, y_pred))
print(metrics.classification_report(Y_test, y_pred))
print(metrics.confusion_matrix(Y_test, y_pred))

Support Vector Machine: 
0.625
              precision    recall  f1-score   support

           0       0.00      0.00      0.00         0
           1       1.00      0.62      0.77       168

    accuracy                           0.62       168
   macro avg       0.50      0.31      0.38       168
weighted avg       1.00      0.62      0.77       168

[[  0   0]
 [ 63 105]]


C:\Users\Nahiyan Bin Noor\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\Nahiyan Bin Noor\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\Nahiyan Bin Noor\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


With SVM the accuracy is 62.5%

# Decision Tree

In [27]:
from sklearn import datasets
from sklearn import metrics
from sklearn.tree import DecisionTreeClassifier
model = DecisionTreeClassifier()
model.fit(X_train, Y_train)
# make predictions
y_pred = model.predict(X_test)
# summarize the fit of the model
print("Classification and Regression Trees")
print(metrics.accuracy_score(Y_test, y_pred))
print(metrics.classification_report(Y_test, y_pred))
print(metrics.confusion_matrix(Y_test, y_pred))

Classification and Regression Trees
0.48214285714285715
              precision    recall  f1-score   support

           0       0.00      0.00      0.00         0
           1       1.00      0.48      0.65       168

    accuracy                           0.48       168
   macro avg       0.50      0.24      0.33       168
weighted avg       1.00      0.48      0.65       168

[[ 0  0]
 [87 81]]


C:\Users\Nahiyan Bin Noor\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\Nahiyan Bin Noor\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\Nahiyan Bin Noor\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


The accuracy with Decision Tree is the lowest and which is 48.21%